aes/src/macros.rs

```diff
@@ -0,0 +1,104 @@
+// TODO: remove this allow after migration of the soft backend
+#[allow(unused_macros)]
+macro_rules! impl_backends {
+    (
+        enc_name = $enc_name:ident,
+        dec_name = $dec_name:ident,
+        key_size = $key_size:ty,
+        keys_ty = $keys_ty:ty,
+        par_size = $par_size:ty,
+        expand_keys = $expand_keys:expr,
+        inv_keys = $inv_keys:expr,
+        encrypt = $encrypt:expr,
+        encrypt_par = $encrypt_par:expr,
+        decrypt = $decrypt:expr,
+        decrypt_par = $decrypt_par:expr,
+) => {
+        #[derive(Clone)]
+        pub(crate) struct $enc_name {
+            keys: $keys_ty,
+        }
+
+        impl cipher::BlockSizeUser for &$enc_name {
+            type BlockSize = cipher::consts::U16;
+        }
+
+        impl cipher::ParBlocksSizeUser for &$enc_name {
+            type ParBlocksSize = $par_size;
+        }
+
+        impl cipher::KeySizeUser for $enc_name {
+            type KeySize = $key_size;
+        }
+
+        impl cipher::KeyInit for $enc_name {
+            #[inline]
+            fn new(key: &cipher::Key<Self>) -> Self {
+                let keys = unsafe { $expand_keys(key.as_ref()) };
+                Self { keys }
+            }
+        }
+
+        impl cipher::BlockBackend for &$enc_name {
```

Comment:
`BlockBackend` is implemented for reference because its methods works with `&mut self`. We probably should introduce two separate traits `BlockCipherBackend` (with `&self` methods) and `BlockModeBackend` (with `&mut self` methods).